### PR TITLE
Making modules reusable outside of Metricbeat

### PR DESCRIPTION
**Pull request by andrewkroh**
TL;DR See the ModuleRunner [example](https://godoc.org/github.com/andrewkroh/beats/metricbeat/beater#example-ModuleRunner).

ModuleWrapper is now exported so that it can be used outside of the
package. ModuleWrapper provides a simple Start() method that outputs
to a golang channel. This makes testing of ModuleWrapper quite easy
because it has few outside dependencies.

Added a facade called ModuleRunner that provides simple Start() and
Stop() methods and publishes to a publisher.Client.

There are godoc examples demonstrating how to use both [ModuleWrapper](https://godoc.org/github.com/andrewkroh/beats/metricbeat/beater#example-ModuleWrapper)
and [ModuleRunner](https://godoc.org/github.com/andrewkroh/beats/metricbeat/beater#example-ModuleRunner).

/cc @vjsamuel

### CODE REVIEW

File: metricbeat/beater/metricbeat.go [hunk 1]
```diff
@@ -5,17 +5,20 @@ import (
...
 
 // Metricbeat implements the Beater interface for metricbeat.
 type Metricbeat struct {
-	done    chan struct{}    // Channel used to initiate shutdown.
-	config  *Config          // Metricbeat specific configuration data.
-	modules []*moduleWrapper // Active list of modules.
+	done      chan struct{}    // Channel used to initiate shutdown.
+	config    *Config          // Metricbeat specific configuration data.
+	modules   []*ModuleWrapper // Active list of modules.
+	pubClient publisher.Client // Publisher client.
```
Comment (ruflin) [inline metricbeat/beater/metricbeat.go]: Should we call this just client identical to the beat-generator? https://github.com/elastic/beat-generator/blob/master/%7B%7Bcookiecutter.beat%7D%7D/beater/%7B%7Bcookiecutter.beat%7D%7D.go#L19

**Comment by andrewkroh** [comment]
jenkins, test it

File: metricbeat/beater/module_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// +build !integration
+
+package beater_test
```
Comment (ruflin) [inline metricbeat/beater/module_test.go]: I assume it is intentional that you use a test package here to have it tested from the outside. Perhaps worth a note?

File: metricbeat/beater/publish.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// PublishChannels publishes the events read from each channel to the given
+// publisher client. If the publisher client blocks for any reason then events
+// will not be read from the given channels.
+//
+// This method blocks until all of the channels have been closed
+// and are fully read. To stop the method immediately, close the channels and
+// close the publisher client to ensure that publishing does not block. This
+// may result is some events being discarded.
+func PublishChannels(client publisher.Client, cs ...<-chan common.MapStr) {
```
Comment (ruflin) [inline metricbeat/beater/publish.go]: Nice. Wasn't aware that also works with channels. One more thing learned.

File: metricbeat/beater/runner.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package beater
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+// ModuleRunner is a facade for a ModuleWrapper that provides a simple interface
```
Comment (ruflin) [inline metricbeat/beater/runner.go]: I would prefer to have all the Wrapper and Runner stuff in a separate package, but didn't come up with a good name yet.

File: metricbeat/beater/module_test.go (hunk 2 above)
Comment (andrewkroh) [inline metricbeat/beater/module_test.go]: Yes, intentional. This is to ensure no unexported functions are required to make use of the modules (black-box testing).

File: metricbeat/beater/runner.go (hunk 4 above)
Comment (andrewkroh) [inline metricbeat/beater/runner.go]: Now that it has grown to be usable outside of Metricbeat's `Beater` implementation, I agree.

File: metricbeat/beater/metricbeat.go (hunk 1 above)
Comment (andrewkroh) [inline metricbeat/beater/metricbeat.go]: Fixed.

File: metricbeat/beater/runner.go (hunk 4 above)
Comment (ruflin) [inline metricbeat/beater/runner.go]: Any idea for package name? Or should we put it into `mb`?
